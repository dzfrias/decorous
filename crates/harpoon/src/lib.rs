mod span;

use arrayvec::ArrayVec;
use itertools::{EitherOrBoth, Itertools};
pub use span::Span;
use std::{collections::VecDeque, ops::Deref, str::Chars};

#[derive(Debug, Clone, PartialEq)]
pub struct Peeked<const SIZE: usize>(ArrayVec<char, SIZE>);

impl<const SIZE: usize> Deref for Peeked<SIZE> {
    type Target = [char];

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[derive(Debug)]
pub struct Harpoon<'a> {
    source: &'a str,
    chars: Chars<'a>,

    peek_buf: VecDeque<char>,
    current: Option<char>,

    idx: usize,
}

impl<'a> Harpoon<'a> {
    pub fn new(input: &'a str) -> Self {
        let harpoon = Self {
            source: input,
            chars: input.chars(),
            current: None,
            peek_buf: VecDeque::new(),
            idx: 0,
        };
        harpoon
    }

    pub fn consume(&mut self) -> Option<char> {
        if let Some(next) = self.peek_buf.pop_front() {
            self.idx += next.len_utf8();
            self.current = Some(next);
            return self.current;
        }

        let next = self.chars.next();
        if let Some(next) = next {
            self.idx += next.len_utf8();
        }
        self.current = next;
        self.current
    }

    pub fn current(&self) -> Option<char> {
        self.current
    }

    pub fn peek(&mut self) -> Option<char> {
        self.peek_n_const::<1>().first().copied()
    }

    pub fn peek_is(&mut self, expected: char) -> bool {
        self.peek().is_some_and(|c| c == expected)
    }

    pub fn peek_is_any(&mut self, expecteds: &str) -> bool {
        self.peek().is_some_and(|c| expecteds.contains(c))
    }

    pub fn peek_n_const<const N: usize>(&mut self) -> Peeked<N> {
        let mut arrvec = ArrayVec::new();
        let remaining = N.saturating_sub(self.peek_buf.len());
        for _ in 0..remaining {
            let Some(next) = self.chars.next() else {
                break;
            };
            self.peek_buf.push_back(next);
        }
        for i in self.peek_buf.iter().take(N) {
            arrvec.push(*i);
        }
        Peeked(arrvec)
    }

    pub fn peek_n(&mut self, n: usize) -> impl Iterator<Item = char> + '_ {
        let remaining = n.saturating_sub(self.peek_buf.len());
        for _ in 0..remaining {
            let Some(next) = self.chars.next() else {
                break;
            };
            self.peek_buf.push_back(next);
        }
        self.peek_buf.iter().take(n).cloned()
    }

    pub fn consume_while<F>(&mut self, mut f: F)
    where
        F: FnMut(char) -> bool,
    {
        while self.peek().is_some_and(&mut f) {
            self.consume();
        }
    }

    pub fn consume_until(&mut self, stopper: char) {
        self.consume_while(|c| c != stopper);
    }

    pub fn try_consume(&mut self, s: &str) -> bool {
        if self.peek_equals(s) {
            self.consume_n(s.chars().count());
            true
        } else {
            false
        }
    }

    pub fn peek_equals(&mut self, s: &str) -> bool {
        !self
            .peek_n(s.len())
            .zip_longest(s.chars())
            .any(|either| !matches!(either, EitherOrBoth::Both(a, b) if a == b))
    }

    pub fn consume_n(&mut self, n: usize) {
        for _ in 0..n {
            self.consume();
        }
    }

    pub fn offset(&self) -> usize {
        self.idx
    }

    pub fn harpoon<F>(&mut self, mut f: F) -> Span<'a>
    where
        F: FnMut(&mut Harpoon),
    {
        let start = self.idx;
        f(self);
        let t = &self.source[start..self.source.len() - self.source[self.idx..].len()];
        Span::new(t, start)
    }

    pub fn source(&self) -> &'a str {
        self.source
    }
}

impl Clone for Harpoon<'_> {
    fn clone(&self) -> Self {
        Self {
            source: self.source().clone(),
            chars: self.source()[self.offset()..].chars(),
            peek_buf: VecDeque::new(),
            current: self.current(),
            idx: self.offset(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn peek_stops_at_length_of_peek_buffer() {
        let mut harpoon = Harpoon::new("1234");
        assert_eq!(Some('1'), harpoon.peek());
        assert_eq!(Some('1'), harpoon.peek());
        harpoon.consume();
        assert_eq!(&['2', '3', '4'], &harpoon.peek_n_const::<3>()[..]);
        assert_eq!(&['2', '3', '4'], &harpoon.peek_n_const::<3>()[..]);
        assert_eq!(&['2', '3', '4'], &harpoon.peek_n_const::<4>()[..]);
    }

    #[test]
    fn harpoon_is_right_exclusive() {
        let mut harpoon = Harpoon::new("1234");
        let span = harpoon.harpoon(|h| {
            assert_eq!(Some('1'), h.consume());
            assert_eq!(Some('2'), h.consume());
        });
        assert_eq!(0, span.start());
        assert_eq!(2, span.end());
        assert_eq!("12", span.text());
    }

    #[test]
    fn consume_while_doesnt_take_failing_char() {
        let mut harpoon = Harpoon::new("1234");
        harpoon.consume_while(|c| c != '4');
        assert_eq!(Some('4'), harpoon.consume());
    }

    #[test]
    fn try_consume_consumes_nothing_if_no_match() {
        let mut harpoon = Harpoon::new("1234");
        assert!(!harpoon.try_consume("33"));
        assert_eq!(Some('1'), harpoon.consume());
    }
}
