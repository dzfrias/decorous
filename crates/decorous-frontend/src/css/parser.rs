use std::borrow::Cow;

use harpoon::Harpoon;
use rslint_parser::AstNode;

use super::{
    ast::{AtRule, Css, Declaration, Pseudo, RegularRule, Rule, Selector, SelectorPart, Value},
    error::{ParseError, ParseErrorType},
};
use crate::{errors::Help, location::Location};

pub type Result<T> = std::result::Result<T, ParseError<Location>>;

#[derive(Debug)]
pub struct Parser<'a> {
    harpoon: Harpoon<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            harpoon: Harpoon::new(source),
        }
    }

    pub fn parse(mut self) -> Result<Css<'a>> {
        let mut rules = vec![];
        self.skip_whitespace();
        while self.harpoon.peek().is_some() {
            rules.push(self.parse_rule()?);
            self.skip_whitespace();
        }
        Ok(Css::new(rules))
    }

    fn parse_rule(&mut self) -> Result<Rule<'a>> {
        if self.harpoon.peek_is('@') {
            return Ok(Rule::At(self.parse_at_rule()?));
        }

        let selector = self.parse_selector()?;
        self.expect_consume('{')?;
        let mut declarations = vec![];
        self.skip_whitespace();
        while !self.harpoon.peek_is('}') && self.harpoon.peek().is_some() {
            declarations.push(self.parse_declaration()?);
            self.skip_whitespace();
        }
        self.expect_consume('}')?;

        Ok(Rule::Regular(RegularRule::new(selector, declarations)))
    }

    fn parse_at_rule(&mut self) -> Result<AtRule<'a>> {
        debug_assert_eq!(Some('@'), self.harpoon.consume());
        let name = self
            .harpoon
            .harpoon(|h| {
                h.consume_while(|c| !c.is_whitespace());
            })
            .text();
        if name.is_empty() {
            return Err(ParseError::new(
                ParseErrorType::ExpectedMediaQueryName,
                Location::from_source(self.harpoon.offset() - 1, self.harpoon.source()),
                None,
            ));
        }
        self.skip_whitespace();
        let additional = self
            .harpoon
            .harpoon(|h| {
                h.consume_while(|c| !matches!(c, '{' | ';'));
            })
            .text();
        if self.harpoon.peek_is(';') {
            debug_assert_eq!(Some(';'), self.harpoon.consume());
            return Ok(AtRule::new(name, additional, None));
        }

        self.expect_consume('{')?;
        self.skip_whitespace();
        let mut rules = vec![];
        while !self.harpoon.peek_is('}') && self.harpoon.peek().is_some() {
            rules.push(self.parse_rule()?);
            self.skip_whitespace();
        }
        self.expect_consume('}')?;

        Ok(AtRule::new(name, additional, Some(rules)))
    }

    fn parse_selector(&mut self) -> Result<Selector<'a>> {
        let mut parts = vec![];
        parts.push(self.parse_selector_part()?);
        self.skip_whitespace();
        while self.harpoon.peek_is(',') {
            debug_assert_eq!(Some(','), self.harpoon.consume());
            self.skip_whitespace();
            parts.push(self.parse_selector_part()?);
            self.skip_whitespace();
        }

        if parts.len() == 1
            && parts
                .first()
                .is_some_and(|p| p.text().is_some_and(|t| t.is_empty()) && p.pseudoes().is_empty())
        {
            return Err(ParseError::new(
                ParseErrorType::ExpectedSelector,
                Location::from_source(self.harpoon.offset(), self.harpoon.source()),
                None,
            ));
        }

        Ok(Selector::new(parts))
    }

    fn parse_selector_part(&mut self) -> Result<SelectorPart<'a>> {
        fn parse_any<'a>(harpoon: &mut Harpoon<'a>) -> &'a str {
            harpoon
                .harpoon(|harpoon| {
                    harpoon.consume_while(|c| !matches!(c, '{' | ':' | ',') && !c.is_whitespace());
                })
                .text()
        }

        let text = if !self.harpoon.peek_is(':') {
            Some(Cow::Borrowed(parse_any(&mut self.harpoon)))
        } else {
            None
        };

        let mut pseudoes = vec![];
        while self.harpoon.peek_is(':') {
            debug_assert_eq!(Some(':'), self.harpoon.consume());
            if self.harpoon.peek_is(':') {
                debug_assert_eq!(Some(':'), self.harpoon.consume());
                pseudoes.push(Pseudo::Element(parse_any(&mut self.harpoon)));
            } else {
                let class_name = self
                    .harpoon
                    .harpoon(|harpoon| {
                        harpoon.consume_while(|c| {
                            !c.is_whitespace() && !matches!(c, '{' | ':' | '(' | ',')
                        });
                    })
                    .text();
                let value = if self.harpoon.peek_is('(') {
                    debug_assert_eq!(Some('('), self.harpoon.consume());
                    let v = self
                        .harpoon
                        .harpoon(|harpoon| harpoon.consume_until(')'))
                        .text();
                    self.expect_consume(')')?;
                    Some(v)
                } else {
                    None
                };
                pseudoes.push(Pseudo::Class {
                    name: class_name,
                    value,
                });
            }
        }

        Ok(SelectorPart::new(text, pseudoes))
    }

    fn parse_declaration(&mut self) -> Result<Declaration<'a>> {
        let name = self
            .harpoon
            .harpoon(|harpoon| harpoon.consume_until(':'))
            .text();
        let line = self.harpoon.line();
        self.expect_consume(':')?;
        self.skip_whitespace();
        let mut values = vec![];
        values.push(self.parse_value()?);
        self.skip_whitespace();
        while !self.harpoon.peek_is_any(";{}:") && self.harpoon.peek().is_some() {
            values.push(self.parse_value()?);
            self.skip_whitespace();
        }
        self.expect_consume_with_help(
            ';',
            Some(Help::with_line(
                line as u32,
                "declaration needs a closing semicolon",
            )),
        )?;
        Ok(Declaration::new(name, values))
    }

    fn expect_consume(&mut self, expected: char) -> Result<()> {
        self.expect_consume_with_help(expected, None)
    }

    fn expect_consume_with_help(&mut self, expected: char, help: Option<Help>) -> Result<()> {
        if self.harpoon.peek_is(expected) {
            self.harpoon.consume();
            Ok(())
        } else {
            Err(ParseError::new(
                ParseErrorType::ExpectedCharacter(expected),
                Location::from_source(self.harpoon.offset(), self.harpoon.source()),
                help,
            ))
        }
    }

    fn parse_value(&mut self) -> Result<Value<'a>> {
        if self.harpoon.peek_is('{') {
            debug_assert_eq!(Some('{'), self.harpoon.consume());
            let offset = self.harpoon.offset();
            let contents = self.harpoon.harpoon(|h| h.consume_until('}')).text();
            self.expect_consume('}')?;
            let res = rslint_parser::parse_expr(contents, 0).ok().map_err(|err| {
                ParseError::new(
                    ParseErrorType::JavaScriptParseError(err),
                    Location::from_source(offset, self.harpoon.source()),
                    None,
                )
            })?;
            Ok(Value::Mustache(res.syntax().clone()))
        } else {
            let t = self
                .harpoon
                .harpoon(|h| h.consume_while(|c| !matches!(c, ';' | '{' | '}' | ':')))
                .text();
            Ok(Value::Css(t))
        }
    }

    fn skip_whitespace(&mut self) {
        self.harpoon.consume_while(|c| c.is_whitespace());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! parser_test {
        ($($input:expr),+) => {
            $(
                let parser = Parser::new($input);
                insta::assert_debug_snapshot!(parser.parse());
             )+
        };
    }

    #[test]
    fn can_parse_basic_rules() {
        parser_test!(
            "p { color: green; }",
            "p { color:green; background: red      ; }",
            "p { color: green; background: red; } h1 { color: red; }"
        );
    }

    #[test]
    fn can_parse_more_complex_selectors() {
        parser_test!(
            "p.green { color: green; }",
            "p.green:hover { color: green; }",
            "p.green:has(h1, h2) { color: green; }",
            "p.green:has(h1, h2):hover { color: green; }",
            "p.green:has(h1, h2):hover::after { color: green; }",
            "p::after { color: green; }",
            "p::after, span.yellow { color: green; }"
        );
    }

    #[test]
    fn can_parse_mustache_tags() {
        parser_test!(
            "p { color: {color}; }",
            "p { color: {color} yellow blue; }",
            "p { width: {w}px; }"
        );
    }

    #[test]
    fn can_parse_more_complex_declaration_values() {
        parser_test!(
            "p { color: rgba(1, 2, 3, 4); }",
            "p { font-family: \"Fira Mono\", monospace; }",
            "p { box-shadow: 3px 3px red, 3px 3px olive; }"
        );
    }

    #[test]
    fn can_parse_at_rules() {
        parser_test!(
            "@import \"style.css\";",
            "@media (hover: hover) { p { color: green; } }",
            "@media (hover: hover) {}"
        );
    }

    #[test]
    fn parser_throws_errors_on_invalid_input() {
        parser_test!(
            "@ \"style.css\";",
            "@media (hover: hover)  p { color: green; } }",
            "p.green { color: green color: red; }",
            "p  color: green; }",
            "p { color: green; ",
            "p { color: {###}; }"
        );
    }
}
