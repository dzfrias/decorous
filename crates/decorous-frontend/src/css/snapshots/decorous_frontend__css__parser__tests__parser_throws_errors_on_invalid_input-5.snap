---
source: crates/decorous-frontend/src/css/parser.rs
expression: parser.parse()
---
Err(
    ParseError {
        fragment: Location {
            offset: 18,
            length: 1,
            line: 1,
            column: 18,
        },
        err_type: ExpectedCharacter(
            '}',
        ),
    },
)
